src/tainting/Taint_lval_env.ml

```diff
@@ -236,9 +298,33 @@ let dumb_find { tainted; cleaned; _ } lval =
         | None -> `None
         | Some taints -> `Tainted taints)
 
-let propagate_from prop_var env = VarMap.find_opt prop_var env.propagated
+let propagate_from prop_var env =
+  let opt_taints = VarMap.find_opt prop_var env.taints_to_propagate in
+  let env =
+    if Option.is_some opt_taints then
+      {
+        env with
+        taints_to_propagate = VarMap.remove prop_var env.taints_to_propagate;
+      }
+    else env
+  in
+  (opt_taints, env)
+
+let pending_propagation env prop_var lval =
```

Comment:
Maybe `env` can be the last argument, unless you had some other thoughts in mind.